### PR TITLE
[Pipelining] Clean up hooks in zero bubble

### PR DESCRIPTION
**Pull request by H-Huang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138720
* #138735
* #138504
* #138119

cc @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/distributed/pipelining/_backward.py [hunk 1]
```diff
@@ -248,6 +254,9 @@ def stage_backward_weight(
             grad_outputs=sum(param_group["grads"], tuple()),
             retain_graph=retain_graph,
         )
+        # release grad memory early after use
+        del param_group["grads"]
```
Comment (wconstab) [inline torch/distributed/pipelining/_backward.py:258]: can you say more about this? Does this meaningfully affect memory usage? What happened if we did not do this delete?

File: torch/distributed/pipelining/_backward.py [hunk 2]
```diff
@@ -210,6 +212,10 @@ def hook(grad_inputs):
     else:
         dinputs = None
 
+    # handles are no longer necessary, clean up for consistency
```
[the remark this answers is not in the document]
Comment (H-Huang) [reply torch/distributed/pipelining/_backward.py:222]: oops, yeah mistyped

File: torch/distributed/pipelining/_backward.py (hunk 1 above)
Comment (wconstab) [reply torch/distributed/pipelining/_backward.py:265]: ok, 'saves a few MB' means it is actually meaningful, even if its small.

**Review by wconstab — APPROVED** [review]

**Pull request closed without merging**